File: Matlab/optim_loop.m
function [mh_err, sim_err, srun] = optim_loop(gam,n,k,ns,opts,s,c0)

mh_err = [];
sim_err = [];
sim_et = [];
mh_et = [];
srun = 0;

for i = 1:ns
    tic()
    x = ones(n,k);
    
    for hh = 2:2:k
        x(:,hh) = gamrnd(10,1,n,1);%wblrnd(5,10,n,1);%chi2rnd(10,n,1);%lognrnd(1,.5,n,1);%
    end
    for hh = 3:2:k
        x(:,3) = gamrnd(5,1,n,1);%wblrnd(5,5,n,1);%chi2rnd(5,n,1);%lognrnd(1,.5,n,1);%
    end
    epsl = normrnd(0,1,n,1);
    
    yp=zeros(n,1);
    for j = 1:k-1
        yp = yp + gam(j+1,1)*(x(:,j+1).^gam(end,1));
    end
    yp = yp.^(1/gam(end,1));
    y = yp + gam(1,1)*ones(n,1) + epsl;
    
    %% Calculate posterior mode and Hessian at mode
    % Constructing the Candidate Density for MH Algorithm
    
    %OLS estimates
    bols = inv(x'*x)*x'*y;
    s2 = (y-x*bols)'*(y-x*bols)/(n-k);
    sse=(n-k)*s2;
    bolscov = s2*inv(x'*x);
    bolssd=zeros(k,1);
    for i = 1:k
        bolssd(i,1)=sqrt(bolscov(i,i));
    end
    
    nparam=k+1;
    parm = ones(nparam,1);
    parm(1:k,1)=bols;
    
    try [bmode,~,~,~,~,hess] = fminunc(@(p)post(p,y,x,n),parm,opts);
        
        try a = chol(inv(hess));
        catch a = ones(4);
        end
        
        if (all(all(a)) ~= 1) && (rcond(inv(hess))>eps)
            err = bmode - gam;
            sim_err = [sim_err err];
            et = toc();
            sim_et = [sim_et et];
            srun = srun+1;
            
            if s == [];
                s = 25000;
            end
            postvar = inv(hess);
            for jj = c0
                ss = 10000;
                [~, ~, acceptancerate, ~] = RWMH(y,x,n,bmode, postvar, ss, jj);
                if acceptancerate > .17 && acceptancerate < 0.25
                    c = jj;
                    break
                end
            end
            [Gsim, ~, ~, et] = RWMH(y,x,n,bmode, postvar, s, c);
            
            err2 = mean(Gsim)' - gam;
            mh_err = [mh_err err2];
            mh_et = [mh_et et];

        end
        clear y
    end
end

mh_err = [mh_err; mh_et];
sim_err = [sim_err; sim_et];

end

File: R/optim_loop.R
optim_loop = function(gam,n,k,ns,s = 25000,c0){
  
  sim_err = NULL;
  sim_et = NULL;
  mh_err = NULL;
  mh_et = NULL;
  srun = 0;
  
  for(i in 1:ns){
    tic = Sys.time();
    x = matrix(1,n,k);
    
    for(hh in seq(from = 2, to = k, by = 2)){
      x[,hh] = rgamma(n, shape = 10, rate = 1); 
    }
    for(hh in seq(from = 3, to = k, by = 2)){
      x[,hh] = rgamma(n, shape = 5, rate = 1);
    }
    epsl = rnorm(n, mean = 0, sd = 1);
    
    yp = matrix(0,n,1);
    for(j in 1:k-1){
      yp = yp + gam[j+1]*(x[,j+1]^gam[1]);
    }
    yp = yp^(1/gam[4]);
    y = yp + gam[1]*matrix(1,n,1) + epsl;
    
    ## OLS estimates
    bols = solve(t(x)%*%x)%*%t(x)%*%y;
    s2 = t(y-x%*%bols)%*%(y-x%*%bols)/(n-k);
    sse=(n-k)%*%s2;
    bolscov = matrix(s2,3,3)*solve(t(x)%*%x);
    bolssd=matrix(0,k,1);
    for(i in 1:k){
      bolssd[i,1]=sqrt(bolscov[i,i]);
    }
    
    nparam=k+1;
    parm = matrix(1,nparam,1);
    parm[1:k,1]=bols;
    
    parm  = as.vector(parm)
    
    res = try(optim(parm,post,x=x,y=y,n=n,hessian=TRUE))
    
    if(class(res) != "try-error"){
      postvar = solve(res$hessian);
      a = try(chol(postvar))
      if(class(a) == "try-error"){
        a = matrix(1,4,4);
      }
      
      if(all(a != 1) & rcond(a) > .Machine$double.eps){
        bmode = res$par;
        err = bmode - gam;
        sim_err = cbind(sim_err, err);
        et = as.numeric(Sys.time() - tic);
        srun = srun+1;
        
        for (jj in c0){
          ss = 10000;
          print(jj)
          part_MH = RWMH(y,x,n,bmode, postvar, ss, jj);
          acceptancerate = part_MH$acceptancerate;
          if(acceptancerate > .17 & acceptancerate < 0.25){
            c = jj;
            break
          }
        } #end for c0
        MH_res = RWMH(y,x,n,bmode, postvar, s, c)
        err2 = colMeans(MH_res$Gsim) - gam;
        mh_err = cbind(mh_err, err2);
        mh_et = cbind(mh_et, MH_res$elapsedtime[[1]]);
        
        
      } # if(all(a != 1) & rcond(a) < .Machine$double.eps)
    } #optim error
    
  } # for ns
  mh_err = rbind(mh_err, mh_et);
  sim_err = rbind(sim_err, sim_et);
  
  return(list("mh_err" = mh_err, "sim_err" = sim_err, "srun" = srun))
  
}
